### PR TITLE
ENH: Improve handling of metric comparison for regression tests

### PR DESCRIPTION
**Pull request by ant0nsc**
- Regression metrics are now read and compared semantically rather than as text files. Hence, formatting of the JSON files no longer matters.
- Fix for bug: if a scalar metric is read from AML before a list metric, the dataframe conversion fails. Hence we can now mix scalar and list metrics without problems.

### CODE REVIEW

File: hi-ml-azure/testazure/testazure/test_azure_util.py [hunk 1]
```diff
@@ -2130,8 +2130,24 @@ def _mock_download_file(child_run_id: str, local_folder_child_run: Path, prefix:
     assert (local_download_folder / str(mock_run_1.id) / remote_file_path).exists()
 
 
-@patch("health_azure.utils.isinstance", return_value=True)
-def test_aggregate_hyperdrive_metrics(_: MagicMock) -> None:
+def test_get_metrics_for_hyperdrive_run() -> None:
+    # test the case where run id is provided
```
Comment (mebristo) [inline hi-ml-azure/testazure/testazure/test_azure_util.py:2134]: Should we also test the case where a run object is provided, and where neither is provided?

File: hi-ml-azure/testazure/testazure/test_azure_util.py [hunk 2]
```diff
@@ -2151,79 +2166,99 @@ def _assert_dataframe_properties(df: pd.DataFrame, num_crossval_splits: int) ->
...
         keep_metrics=invalid_metrics,
-        aml_workspace=ws
     )
-    assert len(metrics_df_3) == 1
-    assert list(metrics_df_3.index) == ['test/accuracy']
-
-    # if a workspace or config file isn't provided, an error should not be raised
-    metrics_df_4 = util.aggregate_hyperdrive_metrics(run=dummy_hyperdrive_run, child_run_arg_name="child_run_index")
-    _assert_dataframe_properties(metrics_df_4, num_crossval_splits)
+    assert len(metrics_df_3) == 2
```
Comment (mebristo) [inline hi-ml-azure/testazure/testazure/test_azure_util.py:2191]: Is metrics_df_3 a dataframe with an empty list for each of the indices listed in invalid metrics? If so could we make that more explicit here, e.g. `assert len(metrics_df_3) == len(invalid_metrics)` and `assert metrics_df_3.loc[invaid_metrics[0]] == []`? Also we could check that the column names match the child_run_index as expected

File: hi-ml/testhiml/testhiml/test_regression_test_utils.py [hunk 3]
```diff
@@ -294,3 +312,280 @@ def test_compare_dictionaries(
...
+    expected = {"a": 1.0}
+    actual = {"a": False}
+    with caplog.at_level(logging.WARNING):
+        assert compare_metrics_dictionaries(expected, actual) == "Mismatch for 1 out of 1 metrics"
+    assert len(caplog.records) == 1
+    assert caplog.messages[0] == "Metric 'a': Actual value has type bool which is not handled."
+
+
+def test_compare_metrics_dictionaries_missing(caplog: pytest.LogCaptureFixture) -> None:
+    """Test for metrics dictionaries where not data is present for an expected metric"""
```
Comment (kenza-bouzid) [inline hi-ml/testhiml/testhiml/test_regression_test_utils.py:386]:
```suggestion
    """Test for metrics dictionaries where no data is present for an expected metric"""
```

File: hi-ml/testhiml/testhiml/test_regression_test_utils.py [hunk 4]
```diff
@@ -294,3 +312,280 @@ def test_compare_dictionaries(
...
+    caplog: pytest.LogCaptureFixture,
+    expected: Dict[str, Any],
+    actual: Dict[str, Any],
+    expected_result: str,
+    messages: List[str],
+) -> None:
+    with caplog.at_level(logging.WARNING):
+        assert _test_compare_metrics_files(tmp_path, expected, actual) == expected_result
+    assert len(caplog.messages) == len(messages), f"Expected {len(messages)} messages, but got {len(caplog.records)}"
+    print(caplog.messages)
```
Comment (kenza-bouzid) [inline hi-ml/testhiml/testhiml/test_regression_test_utils.py:527]: remove print statement?

**Review by kenza-bouzid — APPROVED** [review]
LGTM! wondering if we should allow missing metrics to be part of the dictionary

File: hi-ml-azure/testazure/testazure/test_azure_util.py [hunk 5]
```diff
@@ -2151,79 +2166,99 @@ def _assert_dataframe_properties(df: pd.DataFrame, num_crossval_splits: int) ->
...
+        util.get_metrics_for_run(aml_workspace=ws)
 
     # test the case where we filter a subset of the metrics
     restrict_metrics = ["test/accuracy", "val/loss"]
-    metrics_df_3 = util.get_metrics_for_childless_run(run=dummy_run, keep_metrics=restrict_metrics, aml_workspace=ws)
-    assert len(metrics_df_3) == len(restrict_metrics)
-    expected_metrics_df_3 = expected_metrics_df.loc[restrict_metrics]
-    pd.testing.assert_frame_equal(expected_metrics_df_3, metrics_df_3, check_exact=False, rtol=similarity_tolerance)
+    metrics_3 = util.get_metrics_for_run(run=dummy_run, keep_metrics=restrict_metrics)
+    assert len(metrics_3) == len(restrict_metrics)
```
[the remark this answers is not in the document]
Comment (ant0nsc) [reply hi-ml-azure/testazure/testazure/test_azure_util.py:2217]: refactored into several tests, this is now test a lot better.

File: hi-ml/src/health_ml/utils/regression_test_utils.py [hunk 6]
```diff
@@ -69,6 +66,174 @@ def _check_values_match(expected_v: Any, actual_v: Any, tolerance: float = 1e-5)
                     _check_values_match(expected_value, actual_value, tolerance=tolerance)
 
 
+def _compare_metric_values(expected: Any, actual: Any, tolerance: float = 1e-5) -> str:
+    """Compares an expected and an actual value coming from a metrics dictionary. If the values are numeric
+    and are close enough (within the given tolerance), an empty string is returned. String values
+    are always considered to match, because they usually contain AzureML resource ID. If the actual
```
Comment (mebristo) [inline hi-ml/src/health_ml/utils/regression_test_utils.py:72]: Why is the Azure ML resource ID returned in the metrics? Should we remove this in the function to get metrics?

**Review by mebristo — APPROVED** [review]
Looks good!

File: hi-ml-azure/src/health_azure/utils.py [hunk 7]
```diff
@@ -1965,63 +1964,106 @@ def aggregate_hyperdrive_metrics(
...
+        run_id=run_id,
+        run=run,
+        keep_metrics=keep_metrics,
+        aml_workspace=aml_workspace,
+        workspace_config_path=workspace_config_path,
+    )
+    metrics_swapped = {}
+    for run_tag, run_metrics in metrics.items():
+        for metric_name, metric_value in run_metrics.items():
+            if metric_name not in metrics_swapped:
```
[the remark this answers is not in the document]
Comment (ant0nsc) [reply hi-ml-azure/src/health_azure/utils.py:1978]: This is covered implicitliy in each call to this method where there are more than 1 metric, and we have those tests.

File: hi-ml-azure/testazure/testazure/test_azure_util.py (hunk 1 above)
Comment (ant0nsc) [reply hi-ml-azure/testazure/testazure/test_azure_util.py:2134]: Done with a big refactoring of that test.

File: hi-ml-azure/testazure/testazure/test_azure_util.py (hunk 2 above)
Comment (ant0nsc) [reply hi-ml-azure/testazure/testazure/test_azure_util.py:2191]: done

File: hi-ml/src/health_ml/utils/regression_test_utils.py [hunk 8]
```diff
@@ -69,6 +66,174 @@ def _check_values_match(expected_v: Any, actual_v: Any, tolerance: float = 1e-5)
...
+    :param actual: The actual value
+    :param tolerance: For numerical values, this is the relative tolerance allowed, defaults to 1e-5
+    :return: An empty string if the expected and actual values match, otherwise a string describing the discrepancy.
+    """
+    if type(actual) in [float, int] and type(expected) in [float, int]:
+        if isclose(actual, expected, rel_tol=tolerance):
+            return ""
+        return f"Expected {expected} but got {actual} (allowed tolerance {tolerance})"
+    elif type(actual) is str and type(expected) is str:
+        return ""
```
[the remark this answers is not in the document]
Comment (ant0nsc) [reply hi-ml/src/health_ml/utils/regression_test_utils.py:85]: I've removed strings for now, regression tests only allow numeric values, which I think should be good enought

File: hi-ml/src/health_ml/utils/regression_test_utils.py (hunk 6 above)
Comment (ant0nsc) [reply hi-ml/src/health_ml/utils/regression_test_utils.py:72]: I've removed strings for now, regression tests only allow numeric values, which I think should be good enough

File: hi-ml-azure/testazure/testazure/test_azure_util.py [hunk 9]
```diff
@@ -2151,79 +2166,99 @@ def _assert_dataframe_properties(df: pd.DataFrame, num_crossval_splits: int) ->
...
-    assert len(metrics_df_4) == 1
-    assert list(metrics_df_4.index) == ['test/accuracy']
-
-    # what happens if we dont provide a workspace?
-    metrics_df_5 = util.get_metrics_for_childless_run(run=dummy_run)
-    pd.testing.assert_frame_equal(expected_metrics_df, metrics_df_5, check_exact=False, rtol=similarity_tolerance)
+    metrics_4 = util.get_metrics_for_run(run=dummy_run, keep_metrics=invalid_metrics)
+    assert len(metrics_4) == len(invalid_metrics)
+    assert set(metrics_4.keys()) == set(invalid_metrics)
+    assert metrics_4[invalid_metrics[1]] == []
```
[the remark this answers is not in the document]
Comment (ant0nsc) [reply hi-ml-azure/testazure/testazure/test_azure_util.py]: refactoring after our discussion yesterday, thanks for brinigng that up!

**Pull request merged**